Add read-only organization field to job templates

AlanCoding opened the pull request:
##### SUMMARY
Connect https://github.com/ansible/awx/issues/3903

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
9.1.1
```


##### ADDITIONAL INFORMATION
(draft ATM)

AlanCoding committed:
Initial (editable) pass of adding JT.organization

This is the old version of this feature from 2019
  this allows setting the organization in the data sent
  to the API when creating a JT, and exposes the field
  in the UI as well

Subsequent commit changes the field from editable
  to read-only, but as of this commit, the machinery
  is not hooked up to infer it from project

AlanCoding committed:
Organization on JT as read-only field

Set JT.organization with value from its project

Remove validation requiring JT.organization

Undo some of the additional org definitions in tests

Revert some tests no longer needed for feature

exclude workflow approvals from unified organization field

revert awxkit changes for providing organization

Roll back additional JT creation permission requirement

Fix up more issues by persisting organization field when project is removed

Restrict project org editing, logging, and testing

Grant removed inventory org admin permissions in migration

Add special validate_unique for job templates
  this deals with enforcing name-organization uniqueness

Add back in special message where config is unknown
  when receiving 403 on job relaunch

Fix logical and performance bugs with data migration

within JT.inventory.organization make-permission-explicit migration

remove nested loops so we do .iterator() on JT queryset

in reverse migration, carefully remove execute role on JT
  held by org admins of inventory organization,
  as well as the execute_role holders

Use current state of Role model in logic, with 1 notable exception
  that is used to filter on ancestors
  the ancestor and descentent relationship in the migration model
    is not reliable
  output of this is saved as an integer list to avoid future
    compatibility errors

make the parents rebuilding logic skip over irrelevant models
  this is the largest performance gain for small resource numbers

AlanCoding commented on awx/main/access.py:
```diff
@@ -1465,6 +1463,10 @@ def get_value(Class, field):
 
+        organization = get_value(Organization, 'organization')
+        if (not organization) or (self.user not in organization.job_template_admin_role):
+            return False
```
Marking key point of argument here.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -1408,11 +1402,7 @@ def get_field_from_model_or_attrs(fd):
-        view = self.context.get('view', None)
-        if not organization and not view.request.user.is_superuser:
-            # Only allow super users to create orgless projects
-            raise serializers.ValidationError(_('Organization is missing'))
```
some version of this may need to be put back in

matburt replied:
Is this taken care of anywhere else in here?

jakemcdermott replied:
> some version of this may need to be put back in

Why?

ryanpetrello replied:
Is this comment still applicable?

AlanCoding commented:
In some commit here, I undid work that made the project's organization unchangeable. I think that will get added back in. The only concern with this is this view:

https://github.com/ansible/awx/blob/505dcf9dd262a718b5b4165cb82fa76aaae5e81f/awx/api/views/organization.py#L131

It allows for changing the project's organization outside of the normal project serializer. My proposed solution would be to simply take away the ability to attach/detach with this view. It is always problematic to attach or detach in a reverse ForeignKey sublist (they only really make sense with many-to-many).

AlanCoding commented:
@jlmitch5 this is a new restriction, it might involve some UI change... but maybe not? This is the behavior with the current state of this PR. If this is sufficient, then we're good.

![Screen Shot 2020-01-23 at 7 54 03 AM](https://user-images.githubusercontent.com/1385596/72986197-abdd3f80-3db5-11ea-95b6-9cf97090b7e1.png)

AlanCoding commented on awx/ui/client/src/templates/job_templates/job-template.form.js:
```diff
@@ -43,6 +43,19 @@ function(NotificationsList, i18n) {
+                    dataPlacement: 'right',
+                    ngDisabled: '!(job_template_obj.summary_fields.user_capabilities.edit || canAddJobTemplate)',
+                    awLookupWhen: '(job_template_obj.summary_fields.user_capabilities.edit || canAddJobTemplate)'
+                },
```
oh this probably isn't right anymore

jakemcdermott replied:
@AlanCoding Are you referring to the user capabilities summary field?

jakemcdermott replied:
Given that it is supposed to be read-only and implicitly inherited from the project relationship, should this field _always_ be disabled?

AlanCoding commented on awx/main/fields.py:
```diff
@@ -197,18 +199,28 @@ def update_role_parentage_for_instance(instance):
             cur_role.implicit_parents = new_parents_json
-            cur_role.save()
+            cur_role.save(update_fields=['implicit_parents'])
+    return (parents_added, parents_removed)
```
I understand that this probably doesn't look good. I have tried reverting the changes in this file, and it doesn't work, and I am reminded of why I made the changes.

When the existing method in migrations

https://github.com/ansible/awx/blob/c626f51dae09fb7770bc313f7d7f28e86ef2b9a3/awx/main/migrations/_rbac.py#L38-L50

is called, it only passes in _additions_ and not _removals_. Normally, this is called within signals, where additions and removals are passed on based on what related manager is changed.

Without passing in removals, I can't use that. Well, that's also the only thing that we have for this sort of thing in migration code. If you pull in current models, it doesn't even come close to working because event the schema changes in this PR break the signals / fields logic for implicit role fields. Even if they didn't break here, they would break with future changes.

Yes, it is _extremely_ concerning that our migrations don't seem to re-cache ancestors for removals of permissions. But that method is also probably razor close to OOMing with real databases, and passes around a massive payload in the SQL. Putting in removals of all root nodes to that is also a recipe for breaking things.

So, feeling kind of cornered, I feel like the only logical thing to do is follow the pattern that the signals are following.

matburt replied:
I think that's fine 👍

ryanpetrello commented on awx/main/access.py:
```diff
@@ -1607,42 +1615,19 @@ def filtered_queryset(self):
-        if not data:  # So the browseable API will work
-            return True
-        return self.user.is_superuser
+        raise NotImplementedError('Direct job creation not possible in v2 API')
```
Man, I'm glad we finally did away with this.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -277,7 +277,7 @@ def _copy_table(table, query, path):
                                  JOIN main_job ON main_unifiedjob.id = main_job.unifiedjob_ptr_id
                                  JOIN django_content_type ON main_unifiedjob.polymorphic_ctype_id = django_content_type.id
                                  JOIN main_project ON main_project.unifiedjobtemplate_ptr_id = main_job.project_id
```
Do we still need to include `main_project` in the `JOIN` now?

AlanCoding replied:
Oh! maybe not. I also don't see any reference to `main_job` other than the JOIN specifically for this.

@rooftopcellist do you think it would be safe to remove both of these lines?

```sql
JOIN main_job ON main_unifiedjob.id = main_job.unifiedjob_ptr_id
...
JOIN main_project ON main_project.unifiedjobtemplate_ptr_id = main_job.project_id
```

I suspected that I might be missing something after editing this code.

AlanCoding commented on awx/main/fields.py:
```diff
@@ -256,20 +268,18 @@ def bind_m2m_changed(self, _self, _role_class, cls):
-                continue
-
-            if type(field_name) == bytes:
-                field_name = field_name.decode('utf-8')
```
Background here:

I added a new test to set expectations. We don't allow tuples in these entries (or bytes). The test ensures we don't have that anywhere. If someone adds support for it later, that's fine. But until that happens, I want an alert raised if the entries are not an expected structure, and the _last_ thing I want is a `continue`.

Commit:
add organization to template views in old and new ui

jlmitch5 commented:
UI portion is finished and confirmed with @trahman73 and @tvo318 

Note on project list organization field that it can't be changed if the project is in use

<img width="513" alt="Screen Shot 2020-02-05 at 3 32 55 PM" src="https://user-images.githubusercontent.com/1342624/73882181-9cd3a400-482f-11ea-8601-0596a2c74718.png">

Data list item added to expanded templates list, with link to the org

<img width="811" alt="Screen Shot 2020-02-05 at 3 45 50 PM" src="https://user-images.githubusercontent.com/1342624/73882219-aeb54700-482f-11ea-8986-f02d5a34c42a.png">

Item added to template detail view in ui_next

<img width="1414" alt="Screen Shot 2020-02-05 at 3 50 21 PM" src="https://user-images.githubusercontent.com/1342624/73882256-bc6acc80-482f-11ea-91e9-9e0c72e44e3f.png">

jakemcdermott commented on awx/ui/client/src/templates/job_templates/job-template.form.js:
```diff
@@ -43,6 +43,19 @@ function(NotificationsList, i18n) {
+                    dataTitle: i18n._('Organization'),
+                    dataContainer: 'body',
+                    dataPlacement: 'right',
+                    ngDisabled: '!(job_template_obj.summary_fields.user_capabilities.edit || canAddJobTemplate)',
```
Given that:
- the project organization is read-only when the project is related to at least one job template
- the job template organization is implicitly inherited from the project
- all projects must have an organization
- all job templates must have a project

Shouldn't this field _always_ be disabled?  @AlanCoding @jlmitch5

AlanCoding replied:
I think we just revert this file entirely. It was from the previous version of the feature when it was still editable (and still kind of speculative). I don't think the input box is desired at all.

jakemcdermott replied:
sgtm :+1:

AlanCoding committed:
Various JT.organization cleanup items

cleanup from PR review suggestions

bump migration number

fix test

revert change to old-app JT form no longer needed

AlanCoding commented on awx/main/access.py:
```diff
@@ -1504,22 +1502,31 @@ def can_start(self, obj, validate_license=True):
+        # standard type of check for organization - cannot change the value
+        # unless posessing the respective job_template_admin_role, otherwise non-blocking
+        if not self.check_related('organization', Organization, data, obj=obj, role_field='job_template_admin_role'):
+            return False
```
ugh, should remove this

jakemcdermott reviewed (approved):
ui changes lgtm

AlanCoding commented on awxkit/awxkit/api/resources.py:
```diff
@@ -63,6 +63,7 @@ class Resources(object):
     _inventory_related_root_groups = r'inventories/\d+/root_groups/'
     _inventory_related_script = r'inventories/\d+/script/'
     _inventory_related_update_inventory_sources = r'inventories/\d+/update_inventory_sources/'
+    _inventory_scan_job_templates = r'inventories/\d+/scan_job_templates/'
```
this doesn't look right. Maybe some rebase issue?

matburt commented on awx/main/tests/functional/test_projects.py:
```diff
@@ -213,34 +213,14 @@ def test_project_credential_protection(post, put, project, organization, scm_cre
 
 
-@pytest.mark.django_db()
-def test_create_project_null_organization(post, organization, admin):
```
Why remove these tests? It seems like a valid test to verify.

matburt commented on awx/main/tests/functional/test_rbac_job_start.py:
```diff
@@ -119,10 +121,20 @@ def test_normal_relaunch_via_job_template(self, job_no_prompts, rando):
+        assert job_no_prompts.organization == organization
+        organization.execute_role.members.add(rando)
+        job_no_prompts.job_template.delete()
+        job_no_prompts.job_template = None  # Django should do this for us, but it does not
```
Should it though? 😄

matburt reviewed (commented):
I like this PR, very thorough... I just have a few questions....

jakemcdermott commented on awx/api/generics.py:
```diff
@@ -548,6 +548,15 @@ def get_description_context(self):
 
+    def get_queryset(self):
+        if hasattr(self, 'parent_key'):
+            # Prefer this filtering because ForeignKey allows us more assumptions
```
What assumptions are you referring to?

AlanCoding replied:
We have a reverse ForeignKey relationship available, and that's what this would have historically used. Say you want WFJTs in an organization. `org.workflows.all()`, good. But that reverse relationship has changed, now it's all unified job templates in the organization. You can still use that `org.unifiedjobtemplates.all()`... but that's not accurate for that particular view. That would return a queryset of all UJTs in the org, not what you want for the list of workflows in the org.

I'm leveraging the fact that `WorkflowJobTemplate.objects.filter(organization=org)` works the same now as it did before. And it's arguably better in several ways. You don't introduce the `.distinct()` to the queryset, and it's just a lot cleaner.

Listing by the reverse relationship name is absolutely needed, specifically for many-to-many relationships, which is mainly what this was written for. But if you know it's a reverse FK (which it is if `parent_key` can be set), then I can't think of any reason to favor the more complex filter over this one, speaking in general terms for all views (it also happens to solve my problems here).

jakemcdermott commented on awx/api/serializers.py:
```diff
@@ -347,6 +348,7 @@ def _get_related(self, obj):
 
     def _generate_named_url(self, url_path, obj, node):
         url_units = url_path.split('/')
+        reset_counters()
```
This seems like an odd thing to add, but it's probably just the my unfamiliarity with `settings.NAMED_URL_GRAPH`.

Whats going on here? Why reset these?

ryanpetrello replied:
Yea, I have the same question.

AlanCoding replied:
(just answering from memory for now, can dive deeper if needed)

You know when you're doing some kind of logic on a graph, and you just have to have some state temporarily saved within the operation you're doing. Like, you know you don't need to visit something twice, so you put a `visited` counter on the objects. Well the problem is when those get left lying around because after you completed the loop or whatever they should no longer be relevant.

Due to some combination of the related links and the new middleware shim to support the old named URL, somehow we wound up calling it more than once in a request. That made the graph logic wrong.

This was definitely eyebrow-raising. There was some brittleness here that we never had to deal with before.

ryanpetrello commented on awx/api/views/mixin.py:
```diff
@@ -175,28 +172,18 @@ def get_serializer_context(self, *args, **kwargs):
-        db_results['job_templates_inventory'] = JobTemplate.accessible_objects(
-            self.request.user, 'read_role').values(JT_inventory_reference).annotate(
-            Count(JT_inventory_reference)).order_by(JT_inventory_reference)
+        db_results['job_templates'] = jt_qs.values('organization').annotate(Count('organization')).order_by('organization')
```
Man, the changes to this mixin really hammer home how much simpler this is.

ryanpetrello commented on awx/main/access.py:
```diff
@@ -1435,9 +1435,7 @@ def can_add(self, data):
-            return (
-                Project.accessible_objects(self.user, 'use_role').exists() or
-                Inventory.accessible_objects(self.user, 'use_role').exists())
+            return Organization.accessible_objects(self.user, 'job_template_admin_role').exists()
```
👍

ryanpetrello commented on awx/main/access.py:
```diff
@@ -1607,42 +1615,19 @@ def filtered_queryset(self):
-                if self.user in role:
-                    return True
-        return False
+            Q(organization__in=org_access_qs)).distinct()
```
Again, *woof*.

ryanpetrello commented on awx/main/access.py:
```diff
@@ -1706,10 +1685,16 @@ def get_method_capability(self, method, obj, parent_obj):
+        if obj.job_template and self.user in obj.job_template.admin_role:
             return True
-        return obj.job_template is not None and self.user in obj.job_template.admin_role
+        # If orphaned, allow org JT admins to stop running jobs
```
Thanks for the comments here, this makes sense.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -275,10 +275,8 @@ def _copy_table(table, query, path):
                                  JOIN django_content_type ON main_unifiedjob.polymorphic_ctype_id = django_content_type.id
-                                 JOIN main_project ON main_project.unifiedjobtemplate_ptr_id = main_job.project_id
-                                 JOIN main_organization ON main_organization.id = main_project.organization_id
+                                 JOIN main_organization ON main_organization.id = main_unifiedjob.organization_id
```
👍👍👍

ryanpetrello commented on awx/main/fields.py:
```diff
@@ -140,8 +141,9 @@ def resolve_role_field(obj, field):
-        if not str(type(obj)) == role_cls:
+        # use extremely generous duck typing to accomidate all possible forms
+        # of the model that may be used during various migrations
+        if obj._meta.model_name != 'role' or obj._meta.app_label != 'main':
```
This is a little gross, but I don't really have any better suggestions.  This sort of thing is pretty tricky when dealing with migrating roles themselves.

ryanpetrello commented on awx/main/fields.py:
```diff
@@ -256,20 +267,18 @@ def bind_m2m_changed(self, _self, _role_class, cls):
+            # moved inside a migration, needed for job_template_organization_field
+            # migration in particular
+            # consistency is assured by unit test awx.main.tests.functional
+            field = getattr(cls, field_name, None)
```
👍

ryanpetrello commented on awx/main/middleware.py:
```diff
@@ -192,21 +192,41 @@ def __init__(self, get_response=None):
         if resource in settings.NAMED_URL_MAPPINGS:
-            url_units[4] = self._named_url_to_pk(settings.NAMED_URL_GRAPH[settings.NAMED_URL_MAPPINGS[resource]],
-                                                 url_units[4])
+            url_units[4] = cls._named_url_to_pk(
```
Man, I really wish when we'd written this originally, we'd just used this urlsplit objects' *named attributes* instead of all the indexes.

ryanpetrello commented on awx/main/migrations/0109_v370_job_template_organization_field.py:
```diff
@@ -0,0 +1,81 @@
+        # Re-compute the role parents and ancestors caching
+        migrations.RunPython(rebuild_jt_parents, migrations.RunPython.noop),
+        # for all permissions that will be removed, make them explicit
+        migrations.RunPython(restore_inventory_admins, restore_inventory_admins_backward),
```
Order of operations here makes sense to me, thanks for the helpful comments 👍

ryanpetrello commented on awx/main/migrations/_rbac.py:
```diff
@@ -35,7 +38,186 @@ def create_roles(apps, schema_editor):
                 obj.save()
 
 
+def delete_all_user_roles(apps, schema_editor):
```
After we merge this PR, we should go clean up all these old unnecessary functions from previous RBAC migrations; this one harkens back to Ansible Tower's 3.3.0 timeline, so it's quite old now.

ryanpetrello commented on awx/main/migrations/_rbac.py:
```diff
@@ -35,7 +38,186 @@ def create_roles(apps, schema_editor):
+            # For instance... filters projects used in job template, where that job template
+            # has same id same as UJT from the outer reference (which it does)
+            intermediary_reverse_rel: OuterRef('id')}
+        ).order_by().values_list('organization')[:1]
```
Yep, okay 👍

ryanpetrello commented on awx/main/migrations/_rbac.py:
```diff
@@ -35,7 +38,186 @@ def create_roles(apps, schema_editor):
+    jt_qs = JobTemplate.objects.filter(inventory__isnull=False)
+    jt_qs = jt_qs.exclude(inventory__organization=F('project__organization'))
+    jt_qs = jt_qs.only('id', 'admin_role_id', 'execute_role_id', 'inventory_id')
+    for jt in jt_qs.iterator():
```
👍 `.iterator()`

ryanpetrello commented on awx/main/migrations/0109_v370_job_template_organization_field.py:
```diff
@@ -0,0 +1,81 @@
+        migrations.RenameField(
+            model_name='unifiedjob',
+            old_name='tmp_organization',
+            new_name='organization',
```
Okay, I really like this approach.

ryanpetrello commented on awx/main/models/jobs.py:
```diff
@@ -199,7 +199,7 @@ class JobTemplate(UnifiedJobTemplate, JobOptions, SurveyJobTemplateMixin, Resour
     ]
     FIELDS_TO_DISCARD_AT_COPY = ['vault_credential', 'credential']
-    SOFT_UNIQUE_TOGETHER = [('polymorphic_ctype', 'name')]
+    SOFT_UNIQUE_TOGETHER = [('polymorphic_ctype', 'name', 'organization')]
```
🎉

ryanpetrello reviewed (approved):
Man, @AlanCoding, this PR is seriously a herculean effort.  I don't see anything obviously wrong with it, and while the migration took a lot of concentration to be sure I understand what was going on, it makes sense and matches my expectations about what we said we wanted to change here.

kdelee commented:
@appuk and I are peering very closely at this now

kdelee commented on awx/main/migrations/_rbac.py:
```diff
@@ -35,7 +38,186 @@ def create_roles(apps, schema_editor):
+    jt_qs = jt_qs.only('id', 'admin_role_id', 'execute_role_id', 'inventory_id')
+    for jt in jt_qs.iterator():
+        org = jt.inventory.organization
+        for role_name in ('admin_role', 'execute_role'):
```
@AlanCoding so we discussed this being source of JT Admin of Organization for the Inventory LOSING permissions to JT on upgrade. Do you plan to change this?

AlanCoding replied:
Yes, I hope to get around to a fix maybe mid-day today. The solution is that we need more logic in the loop in the data migration. Right now, it simplistically equates the JT role (like "admin_role") to the org role (like "admin_role"). This needs to be expanded to look at a JT role like "admin_role" and a corresponding list of org roles, like `["admin_role", "job_template_admin_role"]`, and the user possessing either one of those will fit the criteria of interest.

AlanCoding committed:
Grant org JT admin role in migration as well

AlanCoding committed:
Add named URL docs for uniqueness functionality

lcarva commented:
@AlanCoding, I see that this PR adds a new restriction 

> @jlmitch5 this is a new restriction, it might involve some UI change... but maybe not? This is the behavior with the current state of this PR. If this is sufficient, then we're good.
> 
> ![Screen Shot 2020-01-23 at 7 54 03 AM](https://user-images.githubusercontent.com/1385596/72986197-abdd3f80-3db5-11ea-95b6-9cf97090b7e1.png)

@AlanCoding, with this restriction, how do I move a Project with Job Templates to another Organization?

AlanCoding commented:
That's a valid question, you could file an issue for it.

My thoughts are that there's very little interest in revisiting the role parentage issues that we hit before. Maybe we could attack that as a feature which is a subset of the API-backed copy? I don't think that a simple PATCH changing the organization is going to finish cleanly.
